packages/core/saved-objects/core-saved-objects-api-server-internal/src/lib/repository.ts

```diff
@@ -762,6 +772,247 @@ export class SavedObjectsRepository implements ISavedObjectsRepository {
     );
   }
 
+  /**
+   *  {@inheritDoc ISvedObjectsRepository.bulkDelete}
+   */
+  async bulkDelete(
+    objects: SavedObjectsBulkDeleteObject[],
+    options: SavedObjectsBulkDeleteOptions
+  ): Promise<SavedObjectsBulkDeleteResponse> {
+    const { refresh = DEFAULT_REFRESH_SETTING, force } = options;
+    const namespace = normalizeNamespace(options.namespace);
+    let bulkGetRequestIndexCounter = 0;
+
+    const expectedBulkGetAllDocsWithNamespaceChecksFlag =
+      objects.map<BulkDeleteExpectedBulkGetResult>((object) => {
+        const { type, id } = object;
+        if (!this._allowedTypes.includes(type)) {
+          return {
+            tag: 'Left',
+            value: {
+              id,
+              type,
+              error: errorContent(SavedObjectsErrorHelpers.createUnsupportedTypeError(type)),
+            },
+          };
+        }
+        const requiresNamespacesCheck = this._registry.isMultiNamespace(type);
+        return {
+          tag: 'Right',
+          value: {
+            type,
+            id,
+            ...(requiresNamespacesCheck && { esRequestIndex: bulkGetRequestIndexCounter++ }),
+          },
+        };
+      });
+
+    // get multi-namespace saved objects only to minimise the number of docs to fetch for the preflight multinamespace objects.
+    // Create params for an mget request only from docs that need a multi-namespace preflight check
+    const bulkGetMultiNamespaceDocs = expectedBulkGetAllDocsWithNamespaceChecksFlag
+      .filter(isRight)
+      .filter(({ value }) => value.esRequestIndex !== undefined) // only get docs that need multinamespace checks, I don't think we want to do this filtering
```

Comment:
A lot of these comments will not be needed once some parts of the implementation are extracted into preflight checks. I'll clean them up in future commits.